### ISSUE TITLE
contract: "Maximum call stack size exceeded" error on recursive data

### ISSUE THREAD

**Issue by tash-2s** [polkadot-js/api#4008]
I'm writing an ink! smart contract that has a recursive type value. 

When I pass the generated ABI data to `@polkadot/api-contract`, I get a "Maximum call stack size exceeded" error.

How can I use such type of data with polkadot-js?

# Detail
[Here is the minimum reproducible repository](https://github.com/tash-2s/20210930_polkadot_api-contract_error).

## metadata.json
```javascript
{
  // ...
  "types": [
    {
      "def": {
        "variant": {
          "variants": [
            {
              "name": "A"
            },
            {
              "fields": [
                {
                  "type": 1,
                  "typeName": "ink_prelude::boxed::Box<MyEnum>"
                }
              ],
              "name": "B"
            }
          ]
        }
      },
      "path": [
        "recursive",
        "MyEnum"
      ]
    }
  ]
}
```

## Code
```javascript
const code = new CodePromise(
  api,
  readFileSync("./recursive/target/ink/recursive.contract", "utf8")
);
```

## Error
```
Error: createType(Si0Type):: Maximum call stack size exceeded
    at createTypeUnsafe (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/types/create/createType.cjs:68:18)
    at MetaRegistry.createTypeUnsafe (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/types/create/registry.cjs:364:45)
    at MetaRegistry.createType (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/types/create/registry.cjs:356:17)
    at MetaRegistry.value (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/api-contract/MetaRegistry.cjs:74:21)
    at MetaRegistry.value (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/api-contract/MetaRegistry.cjs:208:88)
    at MetaRegistry.value (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/api-contract/MetaRegistry.cjs:93:102)
    at MetaRegistry.getMetaTypeDef (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/api-contract/MetaRegistry.cjs:271:84)
    at /snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/api-contract/MetaRegistry.cjs:142:52
    at Array.map (<anonymous>)
    at Array.map (/snip/20210930_polkadot_api-contract_error/node_modules/@polkadot/types/codec/AbstractArray.cjs:155:27)
```

Thank you.

**Comment by jacogr** [comment]
Related to https://github.com/polkadot-js/apps/issues/6209 (same underlying cause, recursive types)

**Comment by tash-2s** [comment]
Thank you for your comment @jacogr.

Do you think this issue will be addressed soon? If not, is there any workaround I can take?

I believe MetaRegistry needs fix, but there is a big PR that removes MetaRegistry. https://github.com/polkadot-js/api/pull/3988

Seems that the infinite recursive flow is like this.
1. `new Abi`
2. `MetaRegistry` `getMetaTypeDef`
3. `MetaRegistry` `#extract`
4. `MetaRegistry` `#extractVariant`
5. `MetaRegistry` `#extractVariantSub`
6. `MetaRegistry` `#extractFields`
7. `MetaRegistry` `getMetaTypeDef`
    - same as 2

**Comment by jacogr** [comment]
Yes, in the linked PR the old ABI scale-info v0 is replaced with the new scale-info v1. (Actually the above issue may actually already be fixed in there, I will take a look at this ABI as part of the testing for that PR)

**Comment by jacogr** [comment]
Not related to the linked issue above, however this was in the `MetadatRegistry` (which is not present in the new V1-version, i.e. the same issue was found elsewhere and fixed alongside the metadata v14 update). 

TL;DR With the move to #3988 this issue is not present anymore (test-case with the contract added to verify)
